### PR TITLE
Modify the connect/disconnect APIs to clarify their intent and support new requested behavior

### PR DESCRIPTION
**Pull request by rhc54**
Refs https://github.com/pmix/master/pull/69

### CODE REVIEW

**Comment by rhc54** [comment]
No longer of interest

File: RFC0003.md [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+versions are provided. This means that the resource manager should treat the
+failure of any process in the specified group as a reportable event, and take
+appropriate action. Note that different resource managers may respond to
+failures in different manners.
+
+The host RM will assign a unique nspace to the resulting process group. The
+new nspace must be provided when disconnecting from the group. All procs are
+required to disconnect from the nspace prior to terminating.
+
+Note: a process can only engage in _one_ connect operation involving the identical
```
Comment (abouteiller) [inline RFC0003.md]: What's the rationale for this? I would relax it to _one_ connect operation with the same namespace at the same time, or even that namespaces are unique and therefore no restrictions on simultaneity.

File: RFC0003.md [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+```
+PMIX_EXPORT pmix_status_t PMIx_Connect(const pmix_proc_t procs[], size_t nprocs,
+                                       const pmix_info_t info[], size_t ninfo,
+                                       char nspace[], pmix_rank_t *newrank);
+```
+
+The callback function for the non-blocking form of the PMIx\_Connect
+operation will be called once the operation is complete. By default, any participant
+that fails to call "connect" prior to terminating will cause the
+operation to return a "failed" status to all other participants.
```
Comment (abouteiller) [inline RFC0003.md]: the uniform property (i.e. all other participants return failed) can be expensive to achieve. A less costly alternative is to relax to `any participant that fails to call "connect" prior to terminating will cause the operation to return, possibly with a "failed" status, at all other participants. However, there is a loss of functionality in that processes may connect or not non-uniformly which is hard to deal with.  

My take is that we can postpone that issue to later, if the performance gain is seizable. We can then introduce a variant with an INFO key informing that non-uniform return is ok.

File: RFC0003.md (hunk 1 above)
Comment (rhc54) [reply RFC0003.md]: We don't do a collective to assign an identifier to the connect operation, so we use the provided array of proc IDs as the identifier. However, this means that there can only be one active connect operation with that exact array of proc IDs at a time. It doesn't matter if the entire nspace is participating (i.e., a proc ID with PMIX_RANK_WILDCARD), or the array contains specific individual procs. I can adjust the language here to make it clearer if that would help.

File: RFC0003.md (hunk 2 above)
Comment (rhc54) [reply RFC0003.md]: So...are you proposing a change to the language here? Or pointing out a possible future extension?

Comment (jjhursey) [reply RFC0003.md]: I think the suggestion is that we go with the uniform return behavior (as written). Then, possibly, introduce a variant later (say via an info key) that relaxes the return semantics.

Comment (abouteiller) [reply RFC0003.md]: Yes, there's no followup action, if we all agree on that.

Comment (rhc54) [reply RFC0003.md]: Agreed - thanks

**Comment by rhc54** [comment]
@jjhursey @abouteiller It sounds like we are in agreement - can someone formally "second" this submission?

**Comment by jjhursey** [comment]
@abouteiller Had an earlier comment that @rhc54 replied to, but did not follow up on. I'd like to see that resolve before formally seconding this. @abouteiller can you followup on the first comment, and if we are good to go then I'll let you put the "second" flag on this.

**Comment by rhc54** [comment]
Ah, you are correct - I will indeed clarify that language.

File: RFC0003.md (hunk 1 above)
Comment (rhc54) [reply RFC0003.md]: I updated this description - @jjhursey @abouteiller can you please review?

File: RFC0003.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+The host RM will assign a unique nspace to the resulting process group. The
+new nspace must be provided when disconnecting from the group. All procs are
+required to disconnect from the nspace prior to terminating.
+
+Note: any single process _can_ be simultaneously engaged
+in multiple connect operations. For scalability, PMIx does not use a
+collective to assign an identifier to the connect operation. Instead, it
+is expected that the provided array of process IDs will serve to identify
+a specific PMIx\_Connect operation. This means, however, that there can
```
Comment (jjhursey) [inline RFC0003.md]: Does the order of the array distinguish it or is it the set of entries that distinguish it?

Example (nspace=1, {NS,RANK} combos) - are these two the same or different in the eyes of connect:
`array1 = [{1,0},{1,1},{1,2}]`
`array2 = [{1,2},{1,1},{1,0}]`

File: RFC0003.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+only be _one_ active connect operation with that exact array of process
+IDs at a time. For these purposes, an array element consisting of an
+nspace and PMIX\_RANK\_WILDCARD counts as only a single element - i.e.,
+an array that specified every process individually for that nspace would
+result in a different signature than an array containing just the wildcard
+rank entry.
```
Comment (jjhursey) [inline RFC0003.md]: Just to make sure I understand this `PMIX_RANK_WILDCARD` statement correctly.

In the eyes of connect, are these arrays valid, different requests (assume nspace `1` with 2 ranks):
`array1 = [{1, WILD}]`
`array2 = [{1, 0}, {1, WILD}]`
`array3 = [{1, 0}, {1, 1}]`

**Comment by abouteiller** [comment]
As soon as Josh questions are resolved : 👍

File: RFC0003.md (hunk 4 above)
Comment (rhc54) [reply RFC0003.md]: My comment here is actually incorrect, at least per the current implementation. In the prototype branch, we check the array by proc, disregarding order. So your examples (both here and above) would all be considered the same request.

I can correct the comment, or we could decide to make it order-dependent. I'm inclined to leave it as order _independent_ as the alternative feels like we are stepping into specifying implementation.

Thoughts?

Comment (jjhursey) [reply RFC0003.md]: I'm also inclined to keep it order independent. If we find that we need an order-preserving option then we could add an info key for it later - i don't think we need it right now.

**Comment by rhc54** [comment]
I have updated this to reflect the above conversation, and to capture a couple of attributes discussed on the MPI Sessions WG call.

@jjhursey @abouteiller Can you please review?

**Review by abouteiller — APPROVED** [review]

**Comment by rhc54** [comment]
Although we are unaware of anyone who has implemented the server-side backend for PMIx_Connect/Disconnect, we really need to begin enforcing the backward compatibility guarantee. We had changed these signatures to accommodate the work being done in the MPI sessions working group, but we will instead meet that request by adding a set of new APIs (PMIx_Group_*) in the future v4 of the standard. This RFC will therefore be replaced with one for v4 of the standard.

**Pull request merged**